mapping_parenting_tech/analysis/prototyping/clustering_descriptions.py

```diff
@@ -68,34 +76,371 @@
 
 # %%
 # Clustering with hdbscan
+np.random.seed(3667)
 clusterer = hdbscan.HDBSCAN(
-    min_cluster_size=6, min_samples=1, cluster_selection_method="leaf"
+    min_cluster_size=47,
+    min_samples=1,
+    cluster_selection_method="leaf",
+    prediction_data=True,
 )
 clusterer.fit(embedding_clustering)
+logging.info(
+    f"{len(set(clusterer.labels_))} clusters with {sum([bool(x) for x in clusterer.labels_ if int(x) == -1])} unassigned apps"
+)
 
 # %%
 # Prepare dataframe for visualisation
-df = descriptions.copy()
+df = details.copy()
 df["x"] = embedding[:, 0]
 df["y"] = embedding[:, 1]
-df["cluster"] = [str(x) for x in clusterer.labels_]
+df["cluster"] = [int(x) for x in clusterer.labels_]
+df["cluster_confidence"] = [str(x) for x in clusterer.probabilities_]
+
+# %%
+cluster_col = df.columns.get_loc("cluster")
+
+radius = 3
+neigh = KNeighborsClassifier(n_neighbors=radius)
+
+current_unassigned = len(df[df.cluster == -1].cluster.to_list())
+
+cluster_list = df.cluster.to_list()
+
+unassigned_indices = []
+unassigned_indices.extend([i for i, x in enumerate(cluster_list) if x == -1])
+logging.info(f"{len(unassigned_indices)} apps not in clusters")
+
+X_list = [x for i, x in enumerate(embedding_clustering) if i not in unassigned_indices]
+y_list = [i for i in cluster_list if i != -1]
+
+neigh.fit(X_list, y_list)
+
+for x in unassigned_indices:
+    df.iat[x, cluster_col] = neigh.predict([embedding_clustering[x]])
+
+print(f"{len(df[df.cluster == -1].cluster.to_list())} unassigned apps remain.")
 
 # %%
 # Visualise using altair (NB: -1=points haven't been assigned to a cluster)
 fig = (
-    alt.Chart(df, width=500, height=500)
+    alt.Chart(df.reset_index(), width=750, height=750)
     .mark_circle(size=60)
-    .encode(x="x", y="y", tooltip=["cluster", "app_id", "summary"], color="cluster")
+    .encode(x="x", y="y", tooltip=["cluster", "appId", "summary"], color="cluster:N")
 ).interactive()
 
 fig
 
 # %%
-of_interest = ["5", "13"]
-df[df.cluster.isin(of_interest)].app_id.to_list()
+
+fig = (
+    alt.Chart(df.reset_index(), width=750, height=750)
+    .mark_circle(size=60)
+    .encode(
+        x="x",
+        y="y",
+        tooltip=["cluster", "appId", "summary"],
+        color="cluster_relevance:N",
+    )
+).interactive()
```

Comment:
It seems like there is no `cluster_relevance` column. When I ran this cell, there was no colour info in the output (see the attached screenshot). Perhaps you defined it later in the notebook?

![Screenshot 2022-01-28 at 14 16 17](https://user-images.githubusercontent.com/29844253/151562757-6e825960-0f91-48c0-87e2-9715bff93d68.png)